mut/redirects/redirect_main.py

```diff
@@ -0,0 +1,396 @@
+
+def parse_versions(defs):
+    for x in range (0, len(defs)):
+        if defs[x][0] == 'versions':
+            return defs[x][1]
+
+def generate_rule(is_temp, version, old_url, new_url, is_symlink = False):
+
+    global definitions
+    global symlinks
+    global rules
+    
+    context_url = ''
+    
+    for x in range (0, len(definitions)):
+        if (definitions[x][0] == 'base'):
+            context_url = definitions[x][1]
+    
+    # if url contains {version} - substitute in the correct version
+    old_url_sub = rule_substitue(old_url, version)
+    new_url_sub = rule_substitue(new_url, version)
+    
+    # small reformatting the old url
+    if len(old_url_sub) > 0:
+        if old_url_sub[0] != '/':
+            old_url_sub = '/' + old_url_sub
+            
+        if not is_symlink:
+            if old_url_sub[-1] == '/':
+                old_url_sub = old_url_sub[:-1]
+            old_url_sub = '/' + context_url + old_url_sub
+
+    new_rule = Rule()
+    
+    new_rule.is_temp = is_temp
+    new_rule.version = version
+    new_rule.old_url = old_url_sub
+    new_rule.new_url = new_url_sub
+    
+    # check for symlinks
+    if len(symlinks) > 0:
+        for x in range (len(symlinks)):
+            if version == symlinks[x][1]:
+                r = generate_rule(is_temp, symlinks[x][0], old_url, new_url, True)
+                rules.append(r)
+    
+    return new_rule
+
+def rule_substitue(url_string, version):
+    
+    global definitions
+    
+    # look for strings between { }
+    sub_regex = '{(.*?)}'
+    
+    matches = re.findall(sub_regex, url_string, re.DOTALL)
+    
+    if matches:
+        # loop through each match
+        for x in range(0, len(matches)): 
+            # loop through each definition
+            for y in range(0, len(definitions)):
+                # if the match == the definition key
+                if matches[x] == definitions[y][0]:
+                    # substitute the definition value for the match
+                    if type(definitions[y][1]) is str:
+                        url_string = url_string.replace('${' + matches[x] + '}', definitions[y][1])
+                    
+    url_string = url_string.replace('${version}', version)
+    
+    url_string = url_string.strip()
+    
+    return url_string
+    
+def write_to_file(rules, output):
+    
+    # I'm not sure how we want to handle the file name - maybe it makes sense to have it as a parameter?
+    file = open(output + '/htaccess_test.txt', 'w')
+    
+    for rule in rules:
+        line = 'Redirect '
+        
+        if rule.is_temp:
+            line += '302 '
+        else:
+            line += '301 '
+            
+        line += str(rule.old_url) + ' ' + str(rule.new_url)
+        
+        file.write(line)
+        file.write('\n')
+        
+    file.close()
+
+def parse_source_file(source_path, output):
+    
+    line_num = 0
+    
+    with open(source_path) as file:
+        for line in file:
```

Comment:
The `enumerate()` function eliminates the need for the separate `line_num` loop counter you have. 

```python
for line, line_num in enumerate(file):
```